lib/prometheus_exporter/server/puma_collector.rb

```diff
@@ -0,0 +1,48 @@
+module PrometheusExporter::Server
+  class PumaCollector < TypeCollector
+    PUMA_GAUGES = {
+      workers_total: "Number of puma workers.",
+      booted_workers_total: "Number of puma workers booted.",
+      old_workers_total: "Number of old puma workers.",
+      running_threads_total: "Number of puma threads currently running.",
+      request_backlog_total: "Number of requests waiting to be processed by a puma thread.",
+      thread_pool_capacity_total: "Number of puma threads available at current scale.",
+      max_threads_total: "Number of puma threads at available at max scale.",
+    }
+
+    def initialize
+      @puma_metrics = []
+    end
+
+    def type
+      "puma"
+    end
+
+    def metrics
+      return [] if @puma_metrics.length == 0
+
+      metrics = {}
+
+      @puma_metrics.map do |m|
+        labels = {}
+        if m["phase"]
+          labels.merge(phase: m["phase"])
+        end
+
+        PUMA_GAUGES.map do |k, help|
+          k = k.to_s
+          if v = m[k]
+            g = metrics[k] ||= PrometheusExporter::Metric::Gauge.new("puma_#{k}", help)
+            g.observe(v, labels)
+          end
+        end
+      end
+
+      metrics.values
+    end
+
+    def collect(obj)
+      @puma_metrics << obj
```

Comment:
This is an ENORMOUS GIANT issue, it is leaking memory forever and never clearing stuff like process collector does. 

It urgently urgently needs redoing to follow the same pattern in process collector.

Reply:
Sorry about this, if I remember correctly this was based on the process one, which does have slightly different logic, where it cleans up metrics:
```
      now = ::Process.clock_gettime(::Process::CLOCK_MONOTONIC)

      obj["created_at"] = now

      @process_metrics.delete_if do |current|
        obj["pid"] == current["pid"] || (current["created_at"] + MAX_PROCESS_METRIC_AGE < now)
      end
      @process_metrics << obj
```

I think I missunderstood the impact of this at the time and we actually ran into this memory leak at work. There was someone who fixed it, but unfortunately never contributed it, perhaps due to confusion when our company dropped support for ruby. Once again sorry for that.

This is what our current implementation looks like:
```ruby
# frozen_string_literal: true

# collects stats from puma
module PrometheusExporter
  module Instrumentation
    class Puma
      def self.start(client: nil, frequency: 30)
        puma_collector = new
        client ||= PrometheusExporter::Client.default
        Thread.new do
          loop do
            begin
              metric = puma_collector.collect
              client.send_json metric
            rescue StandardError => e
              STDERR.puts("Prometheus Exporter Failed To Collect Puma Stats #{e}")
            ensure
              sleep frequency
            end
          end
        end
      end

      def collect
        metric = {}
        metric[:type] = 'puma'
        collect_puma_stats(metric)
        metric
      end

      def collect_puma_stats(metric)
        stats = JSON.parse(::Puma.stats)

        if stats.key? 'workers'
          metric[:phase] = stats['phase']
          metric[:workers_total] = stats['workers']
          metric[:booted_workers_total] = stats['booted_workers']
          metric[:old_workers_total] = stats['old_workers']

          stats['worker_status'].each do |worker|
            collect_worker_status(metric, worker['last_status'])
          end
        else
          collect_worker_status(metric, stats)
        end
      end

      private

      def collect_worker_status(metric, status)
        return if status.empty?

        metric[:request_backlog_total] ||= 0
        metric[:running_threads_total] ||= 0
        metric[:thread_pool_capacity_total] ||= 0
        metric[:max_threads_total] ||= 0

        metric[:request_backlog_total] += status["backlog"]
        metric[:running_threads_total] += status["running"]
        metric[:thread_pool_capacity_total] += status["pool_capacity"]
        metric[:max_threads_total] += status["max_threads"]
      end
    end
  end
end
```

I see there is another fix more along the lines of the process one, but I could update with this kind of solution if you think it is better.

Once again sorry for the inconvenience.

Reply:
No worries, these kinds of things happen!

Upgrading the puma collector is less urgent now, but I think it might be a nice PR? We can compare to the old one and see if it's more valuable data (I suspect it is.)